File: src/game/production/inventory.ts
import type { PlayerSave } from '../save/types'
import { isAtKitchen, isKitchenAmbientIngredient, type RecipeIngredient } from './recipes'

export function removeIngredients(
  save: PlayerSave,
  ingredients: RecipeIngredient[],
  crafts: number = 1,
): PlayerSave | null {
  if (crafts <= 0) return save
  let inventory = save.inventory.map((stack) => ({ ...stack }))
  for (const ingredient of ingredients) {
    if (isKitchenAmbientIngredient(ingredient.itemId) && isAtKitchen(save)) continue
    const need = ingredient.quantity * crafts
    const stack = inventory.find((entry) => entry.itemId === ingredient.itemId)
    if (!stack || stack.quantity < need) return null
    stack.quantity -= need
  }
  inventory = inventory.filter((stack) => stack.quantity > 0)
  return { ...save, inventory }
}

File: packages/ik_rules/lib/src/production/inventory.dart
import '../save/generated/save_models.dart';
import 'recipes.dart';

/// Spends ingredients for `crafts` repetitions, or null when any are missing.
///
/// Emptied stacks leave the bag, which is why callers must re-resolve inventory
/// indexes afterwards.
PlayerSave? removeIngredients(
  PlayerSave save,
  List<RecipeIngredient> ingredients, [
  num crafts = 1,
]) {
  if (crafts <= 0) return save;
  final inventory = [...save.inventory];
  for (final ingredient in ingredients) {
    final need = ingredient.quantity * crafts;
    final index = inventory.indexWhere((entry) => entry.itemId == ingredient.itemId);
    if (index < 0 || inventory[index].quantity < need) return null;
    inventory[index] = inventory[index].copyWith(quantity: inventory[index].quantity - need);
  }
  return save.copyWith(inventory: inventory.where((stack) => stack.quantity > 0).toList());
}
